### PR TITLE
Geo importer

### PR DESCRIPTION
**Pull request by CPWstatic**

### CODE REVIEW

File: tools/importer/src/main/java/com/vesoft/nebula/tools/Constant.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package com.vesoft.nebula.tools;
+
+public class Constant {
```
Comment (darionyaphet) [inline tools/importer/src/main/java/com/vesoft/nebula/tools/Constant.java]: Nice work . Please replace with interface and implement it :)

File: tools/importer/src/main/java/com/vesoft/nebula/tools/ClientManager.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+import com.vesoft.nebula.graph.client.GraphClient;
+import com.vesoft.nebula.graph.client.GraphClientImpl;
+import org.apache.log4j.Logger;
+
+import java.util.List;
+
+public class ClientManager {
+    private static final Logger LOGGER = Logger.getLogger(ClientManager.class.getClass());
+
+    private static ThreadLocal<GraphClient> clientThreadLocal = new ThreadLocal<>();
```
Comment (darionyaphet) [inline tools/importer/src/main/java/com/vesoft/nebula/tools/ClientManager.java]: So the threads will share the one GraphClient？ How to make sure thread safe？

Comment (CPWstatic) [reply tools/importer/src/main/java/com/vesoft/nebula/tools/ClientManager.java]: **ThreadLocal** makes sure that each thread in threadpool will hold unique instance.

File: tools/importer/src/main/java/com/vesoft/nebula/tools/GeoImporter.java [hunk 3]
```diff
@@ -91,7 +91,7 @@ public void runMultiJob(Options options) throws Exception {
         this.hostAndPorts = options.getHostPort();
 
         readContent();
-        FileWriter fileWriter = new FileWriter(options.errorPath.toFile());
+        FileWriter fileWriter = new FileWriter(options.errorPath.toFile(), true);
```
Comment (darionyaphet) [inline tools/importer/src/main/java/com/vesoft/nebula/tools/GeoImporter.java]: `FileWriter(String fileName, boolean append)` when the file have existed we should not append to the end  ?

Comment (CPWstatic) [reply tools/importer/src/main/java/com/vesoft/nebula/tools/GeoImporter.java]: We will not run if the file exist.

**Comment by CPWstatic** [comment]
> Why don't merge `GeoImporter` and `Importer` into one ?
> 
> IMO, support a flag to control whether to loading the `GEO` is reasonable.

You're right. I'm working on this.

File: tools/importer/src/main/java/com/vesoft/nebula/tools/Importer.java [hunk 4]
```diff
@@ -4,352 +4,251 @@
...
-            return;
+        private List<String> buildGeoEdgeKey(List<String> record) {
+            double lat = Double.parseDouble(record.get(0));
+            double lng = Double.parseDouble(record.get(1));
+            long dstId = Long.parseLong(record.get(2));
+
+            List<Long> cells = indexCells(lat, lng);
+            ArrayList<String> geoKeys = new ArrayList<>();
+            for (Long cellId : cells) {
+                geoKeys.add(String.format(Constant.INSERT_EDGE_VALUE_TEMPLATE, cellId, dstId, 0, ""));
```
Comment (yixinglu) [inline tools/importer/src/main/java/com/vesoft/nebula/tools/Importer.java]: Replace hard code rank 0 with `DEFAULT_GEO_EDGE_RANK`

File: tools/importer/src/main/java/com/vesoft/nebula/tools/Importer.java [hunk 5]
```diff
@@ -4,352 +4,251 @@
...
+                        options.type, options.schemaName, options.columns,
+                        Joiner.on(", ").join(values));
+                LOGGER.debug(String.format("Execute: %s", exec));
+                retCode = client.execute(exec);
+                if (retCode != 0) {
+                    synchronized (csvPrinter) {
+                        csvPrinter.printRecords(records);
                     }
+                    LOGGER.info(String.format("Insert batch failed: %d, cost %d ms",
+                            records.size(), System.currentTimeMillis() - startTime));
```
Comment (yixinglu) [inline tools/importer/src/main/java/com/vesoft/nebula/tools/Importer.java]: It's better to print error message and code of this execution to log file.

File: tools/importer/src/main/java/com/vesoft/nebula/tools/Options.java [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+import java.nio.file.Paths;
+import java.util.List;
+
+import org.kohsuke.args4j.Option;
+
+
+public class Options {
+    @Option(name = "-a", aliases = "--address", usage = "thrift service addresses. ip:port",
+            required = true)
+    String addresses;
```
Comment (yixinglu) [inline tools/importer/src/main/java/com/vesoft/nebula/tools/Options.java]: We can set some default values for user to simplify usage in command line. for example:
- address=127.0.0.1:3699; 
- user=user; 
- password=password;

File: tools/importer/src/main/java/com/vesoft/nebula/tools/reader/CsvReader.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+        Path path = Paths.get(filePath);
+        if (!Files.exists(path) || !Files.isReadable(path)) {
+            throw new IllegalArgumentException(filePath + " is not exist or not readable");
+        }
+        LOGGER.info(String.format("Reading Local FileSystem: %s", filePath));
+        this.csvParser = CSVParser.parse(
+                new File(filePath),
+                Charset.forName("UTF-8"),
+                CSVFormat.DEFAULT
+                        .withFirstRecordAsHeader()
```
Comment (yixinglu) [inline tools/importer/src/main/java/com/vesoft/nebula/tools/reader/CsvReader.java]: Should make user to configure whether the **first record line** is header

File: tools/importer/src/main/java/com/vesoft/nebula/tools/reader/CsvReader.java [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+        Configuration configuration = new Configuration();
+        FileSystem fileSystem = FileSystem.get(URI.create(filePath), configuration);
+        LOGGER.info(String.format("Reading HDFS: %s", filePath));
+        try (InputStream inputStream = fileSystem.open(
+                new org.apache.hadoop.fs.Path(filePath))) {
+            this.csvParser = CSVParser.parse(
+                    inputStream,
+                    Charset.forName("UTF-8"),
+                    CSVFormat.DEFAULT
+                            .withFirstRecordAsHeader()
```
Comment (yixinglu) [inline tools/importer/src/main/java/com/vesoft/nebula/tools/reader/CsvReader.java]: ditto

**Comment by darionyaphet** [comment]
It's looks like we are going to use `golang` to implement `Importer` case.

Please hold on this one & thanks for your great work. I really like it  >-<

File: tools/importer/src/main/java/com/vesoft/nebula/tools/reader/CsvReader.java (hunk 7 above)
Comment (CPWstatic) [reply tools/importer/src/main/java/com/vesoft/nebula/tools/reader/CsvReader.java]: The format seems different when the csv file is dump from mysql/oracle/mongo, we need to support them. We should enhance it when we implement the go-importer.

**Pull request closed without merging**